### ISSUE TITLE
Virtual keyboard showing up always

### ISSUE THREAD

**Issue by jelhan** [select2/select2#1541]
If select2 is used the virtual keyboard is showing up always if a user opening or closing an select.

It's caused at least by two reasons: 
a) focus is called on search input field on opening select (line 1785 in select2-3.4.1)
b) focus is called on focusser input field on close (line 1808), focus (line 1817), cancel (line 1830)

The virtual keyboard is also showing up if search is disabled by minimumResultsForSearch. So users of mobile devices get virtual keyboard but can't input something. Also virtual keyboard is overlaying select options. So select is nearly not usable on mobile device.

**Comment by Xcellere** [comment]
Was this ever resolved in the 4.0 beta?

**Comment by kevin-brown** [comment]
@Xcellere In 4.0 we removed the use of the special hidden input for handling focus (instead just using the default element focus), so the issues here should still be resolved, just in a different way.

Have you run into any issues in the 4.0 beta as far as the virtual keyboard that we should look into?

**Comment by Xcellere** [comment]
Was just curious regarding 3.5.2 and below, haven't gotten 4.0 working yet per the language issue I opened.

**Comment by LvLynx** [comment]
@kevin-brown To check if this is fixed I switched to 4.0.0 beta 2. It`s still not working - virtual keyboard appears, because of search bar auto focus.

Latest chrome 40.0.0214
Android 4.2.2 tablet

Code used to create select
```
<select id="sort_category" name="sort_category" style="width: 300px;">
 <option value="0">Show all</option>
</select>
<script type="text/javascript">
  $("#sort_category").select2({
   data : [...]
  }).val("88").trigger("change");
</script>
```

**Comment by 25prathamesh** [comment]
i am getting same issue on chrome 40.0.22 , android 4.2.2 , samsung galaxy grand 
any solution or hack ?

**Comment by cdransf** [comment]
This worked for me:
```
.select2-focusser,
.select2-search,
.select2-search input {
    display: none !important;
}
```

**Comment by bomba1990** [comment]
I found a  solution to this issue in Version: 3.5.1 Timestamp: Tue Jul 22 18:58:56 EDT 2014

In the function of opening the select2, its a line that call search focus().  I make a check in this line, to evaluate where is a mobile?

https://github.com/select2/select2/blob/stable/3.5/select2.js#L2034

There is mi code:

```javascript
var check = false;
(function(a,b){if(/(android|bb\d+|meego).+mobile|avantgo|bada\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od)|iris|kindle|lge |maemo|midp|mmp|mobile.+firefox|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up\.(browser|link)|vodafone|wap|windows ce|xda|xiino/i.test(a)||/1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac|ac(er|oo|s\-)|ai(ko|rn)|al(av|ca|co)|amoi|an(ex|ny|yw)|aptu|ar(ch|go)|as(te|us)|attw|au(di|\-m|r |s )|avan|be(ck|ll|nq)|bi(lb|rd)|bl(ac|az)|br(e|v)w|bumb|bw\-(n|u)|c55\/|capi|ccwa|cdm\-|cell|chtm|cldc|cmd\-|co(mp|nd)|craw|da(it|ll|ng)|dbte|dc\-s|devi|dica|dmob|do(c|p)o|ds(12|\-d)|el(49|ai)|em(l2|ul)|er(ic|k0)|esl8|ez([4-7]0|os|wa|ze)|fetc|fly(\-|_)|g1 u|g560|gene|gf\-5|g\-mo|go(\.w|od)|gr(ad|un)|haie|hcit|hd\-(m|p|t)|hei\-|hi(pt|ta)|hp( i|ip)|hs\-c|ht(c(\-| |_|a|g|p|s|t)|tp)|hu(aw|tc)|i\-(20|go|ma)|i230|iac( |\-|\/)|ibro|idea|ig01|ikom|im1k|inno|ipaq|iris|ja(t|v)a|jbro|jemu|jigs|kddi|keji|kgt( |\/)|klon|kpt |kwc\-|kyo(c|k)|le(no|xi)|lg( g|\/(k|l|u)|50|54|\-[a-w])|libw|lynx|m1\-w|m3ga|m50\/|ma(te|ui|xo)|mc(01|21|ca)|m\-cr|me(rc|ri)|mi(o8|oa|ts)|mmef|mo(01|02|bi|de|do|t(\-| |o|v)|zz)|mt(50|p1|v )|mwbp|mywa|n10[0-2]|n20[2-3]|n30(0|2)|n50(0|2|5)|n7(0(0|1)|10)|ne((c|m)\-|on|tf|wf|wg|wt)|nok(6|i)|nzph|o2im|op(ti|wv)|oran|owg1|p800|pan(a|d|t)|pdxg|pg(13|\-([1-8]|c))|phil|pire|pl(ay|uc)|pn\-2|po(ck|rt|se)|prox|psio|pt\-g|qa\-a|qc(07|12|21|32|60|\-[2-7]|i\-)|qtek|r380|r600|raks|rim9|ro(ve|zo)|s55\/|sa(ge|ma|mm|ms|ny|va)|sc(01|h\-|oo|p\-)|sdk\/|se(c(\-|0|1)|47|mc|nd|ri)|sgh\-|shar|sie(\-|m)|sk\-0|sl(45|id)|sm(al|ar|b3|it|t5)|so(ft|ny)|sp(01|h\-|v\-|v )|sy(01|mb)|t2(18|50)|t6(00|10|18)|ta(gt|lk)|tcl\-|tdg\-|tel(i|m)|tim\-|t\-mo|to(pl|sh)|ts(70|m\-|m3|m5)|tx\-9|up(\.b|g1|si)|utst|v400|v750|veri|vi(rg|te)|vk(40|5[0-3]|\-v)|vm40|voda|vulc|vx(52|53|60|61|70|80|81|83|85|98)|w3c(\-| )|webc|whit|wi(g |nc|nw)|wmlb|wonu|x700|yas\-|your|zeto|zte\-/i.test(a.substr(0,4)))check = true})(navigator.userAgent||navigator.vendor||window.opera);
if(check == false)
    this.search.focus();
```

**Comment by almogdesign** [comment]
Hi Everyone, 
I would like to still show the search option, but not have the virtual keyboard open automatically on mobile (android) only when the user selects the search box. Is there a way to do this?

**Comment by sandermarechal** [comment]
I would like to know that too. I want the search box to appear but not be focused automatically.

**Comment by kylerbeetun** [comment]
+1 for @bomba1990 idea. Would be great having it as an option.

**Comment by furyscript** [comment]
Bug solution??
I still have bug that when i click on select2 element and I've set minimumInputLength: -1, it's appears the keyboard on mobile phone!

**Comment by karlazz** [comment]
This worked for me:
jQuery('#research .filter').on('select2:open', function (e) {
       jQuery('.select2-search input').prop('focus',false);
});
Substitute your own selector for the first.

**Comment by chinsa** [comment]
This works for me:
$(selector).select2({			
  minimumResultsForSearch: Infinity,
  ...
});
$(".select2-search, .select2-focusser").remove();

**Comment by faiq** [comment]
The negative value is not a complete fix, because if you unfocus and then select another value you will still filter. 

If anyone does not want any filtering at all, you can replace the match function to always return true. 

```
$(selector)select2({
            minimumResultsForSearch: 100,
            formatResult: format_timeframe,
            formatSelection: format_timeframe,
            escapeMarkup: function(m) { return m; },
            matcher: function () {
                return true;
            }
        }).on('select2-opening', function () {
            $(this).siblings('.select2-container').find('.select2-search').remove()
        })
```

**Comment by borys-p** [comment]
+1

**Comment by yuracher** [comment]
+1
faiq, your method working only after open. After closing keyboard showing again.
P.S.: sorry for my English ;)

**Comment by Inggo-inHomePortal** [comment]
+1

**Comment by katiewelton** [comment]
+1 for @almogdesign

**Comment by choma** [comment]
+1 for @karlazz solution. It is the way to go if you want to keep the search box, avoiding the keyboard to pop out until the user click the input.

**Comment by eshem** [comment]
+1 for @karlazz solution.

If you are using an older version of select2, you might want to try ```'select2-open'``` instead of ```'select2:open'```

**Comment by janniconl** [comment]
````
$('#select2').on('select2-focus', function (e) {
    $('.select2-search input').prop('focus',false).removeClass('select2-focused');
});
````

@karlazz his solution somehow did not work for me on Android (6.0) devices. Didnt try on other OSs. The solution above is not the prettiest but did it for me!

**Comment by janniconl** [comment]
````
$('#select2').on('select2-focus', function (e) {
    $('.select2-search input').prop('focus',false).removeClass('select2-focused');
});
````

@karlazz his solution somehow did not work for me on Android (6.0) devices. Didnt try on other OSs. The solution above is not the prettiest but did it for me!

**Comment by amitailanciano** [comment]
+1

**Comment by hheimbuerger** [comment]
I still have the virtual keyboard pop up in multiselection boxes both on select2 v4.0.0 as well as v4.0.1rc1.

**Comment by agliottone** [comment]
my workaround: work in select2 3.5+
include hammerjs and hammerjs jquery plugin

and add this js:
```javascript
$(document).ready(function () {
	if (/Android|webOS|iPhone|iPad|iPod|BlackBerry|BB|PlayBook|IEMobile|Windows Phone|Kindle|Silk|Opera Mini/i.test(navigator.userAgent)) {

		setInterval(function(){
			if($(".select2-results:visible").length){
				if($('body').css('overflow')!="hidden"){
					$("#select2-drop-mask").unbind().off();
					$("#select2-drop-mask").hammer({}).bind("tap", function(e){
						setTimeout(function(){
						$(".select2-dropdown-open").next().select2("close");
						},50)
					    e.preventDefault();
					    return false
					});
					$('body, html').css('overflow','hidden')
				}
			}else{
				$("#select2-drop-mask").unbind();
				$('body, html').css('overflow','auto')
			}
		},500)
	}
})
```

and this css:

```css
#s2id_(selectid) .select2-focusser,
#s2id_(selectid) .select2-search,
#s2id_(selectid) .select2-search input
{
    display: none !important;
}
```

**Comment by isbkch** [comment]
+1

**Comment by raldred** [comment]
I've read this thread, I don't understand why it's closed. workarounds proposed but it seems they don't work for everyone. But it should remain open so someone can actually find a real solution, yo-yo opening and closing isn't helping anyone. People just open duplicates. This is more than an annoyance, UX is incredibly important. Let's work on finding a real solution.

**Comment by peterbunny** [comment]
+1

**Comment by Booyanach** [comment]
+1 same problem

**Comment by Dazag** [comment]
it's kind of weird, noone has done anything to solve this "HUGE" problem... who does want a jquery plugin that is not working fine on mobiles?

**Comment by Booyanach** [comment]
We sort of solved this by adding a button with an x to the right of the input in an angular2 implementation of this...
obviously not the best option but it works...

a work around might be to have a see-through no-text button that is the size of the window, so when people touch outside of the input they're actually touching the button.

**Comment by bdnorris** [comment]
+1 same problem

**Comment by brunobruno** [comment]
on change give focus to another element with tabindex="0"

<div id="container" tabindex="0"></div>
  onchange event:
$('#container').focus();

solved for me

**Comment by Booyanach** [comment]
@brunobruno that goes around the issue that this lib doesn't handle this problem by default tho.

**Comment by giupenni** [comment]
+1 same problem.

Incredible that this issue is still open.

**Comment by arrancscott** [comment]
Leaving my fix for someone else trawling through this page looking for a resolution.

I had my select tag wrapped in a label tag.

Basic example:

`<label>`
  `<select id="select2"></select2>`
`</label>`

Taking this out of the label tags and placing it in a div tag fixed the issue.

**Comment by lenamtl** [comment]
+1

**Comment by Studioartcz** [comment]
WTF, i need search input on phones, how can i fix it?

**Comment by DiscoTim** [comment]
seems to work on android/chrome haven't tested anything else
```
$('select').each(function () {
  $(this).on('select2:open', function(e) {
    $('.select2-search input').blur();
  }).on('select2:close', function(e) {
    $('.select2-search input').blur();
  })
});
```

**Comment by ambroiseRabier** [comment]
i used anilastrio solution : 

```
if (/Android|webOS|iPhone|iPad|iPod|BlackBerry|BB|PlayBook|IEMobile|Windows Phone|Kindle|Silk|Opera Mini/i.test(navigator.userAgent)) {
    $(".select2-search__field").attr("readonly", true);
}
```

**Comment by aldarund** [comment]
Why is this closed ???

**Comment by Studioartcz** [comment]
Is it oficiall fixed?

**Comment by bIbI4k0** [comment]
This bug is also seen in version 4.0.3. Will be official solution? Thank you.

**Comment by jpearceadditions** [comment]
How can this have been such a fundamental bug for years and have been closed and still not be working after major releases?

**Comment by kevin-brown** [comment]
I'm actually going to lock this discussion, as at this point we've moved away from what the original issue was.

1. The virtual keyboard shows up when closing the select
2. The virtual keyboard shows up when opening the select, even when `minimumResultsForSearch` is set

Both of these issues have since been fixed. With that being said, Select2 will still trigger the virtual keyboard when the natural next option is to perform a search. There is no longer a hidden input field that triggers the virtual keyboard, that was removed in Select2 4.0.

_If you have some other issue that occurs in the latest version of Select2, feel free to create a new ticket about it and reference this ticket._
